File: MyMethod/KMeans.java
package KMeans.MyMethod;

import java.io.*;
import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * KMeans算法的实现
 * 包装成工具类，直接传参数调用即可使用
 */

public class KMeans {
    //聚类的个数
    int cluterNum;
    //数据集中的点
    List<Point> points = new ArrayList<>();
    //簇的中心点
    List<Point> centerPoints = new ArrayList<>();
    //聚类结果的集合簇，key为聚类中心点在centerPoints中的下标，value为该类簇下的数据点
    HashMap<Integer, List<Point>> clusters = new HashMap<>();

    public KMeans(String path, int cluterNum){
        this.cluterNum = cluterNum;
        loadData(path);
    }

    //加载数据集
    public void loadData(String path) {
        File file = new File(path);
        try {
            BufferedReader br = new BufferedReader(new FileReader(file));
            String line;
            while ((line = br.readLine()) != null) {
                String[] strs = line.split(" ");
                points.add(new Point(strs[0], strs[1]));
            }
            br.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        //初始化KMeans模型，这里选数据集前classNum个点作为初始中心点
        for (int i = 0; i < cluterNum; i++) {
            centerPoints.add(points.get(i));
            clusters.put(i, new ArrayList<>());
        }
    }

    //KMeans聚类
    public void doKMeans(){
        double err = Integer.MAX_VALUE;
        while (err > 0.01 * cluterNum){
            //每次聚类前清空原聚类结果的list
            for (int key : clusters.keySet()){
                List<Point> list = clusters.get(key);
                list.clear();
                clusters.put(key, list);
            }
            //计算每个点所属类簇
            for (int i=0; i<points.size(); i++){
                dispatchPointToCluster(points.get(i), centerPoints);
            }
            //计算每个簇的中心点，并得到中心点偏移误差
            err = getClusterCenterPoint(centerPoints, clusters);
            show(centerPoints, clusters);
            System.out.println("*************************");
        }
    }

    //计算点对应的中心点，并将该点划分到距离最近的中心点的簇中
    public void dispatchPointToCluster(Point point, List<Point> centerPoints){
        int index = 0;
        double tmpMinDistance = Double.MAX_VALUE;
        for (int i=0; i<centerPoints.size(); i++){
            double distance = calDistance(point, centerPoints.get(i));
            if (distance < tmpMinDistance){
                tmpMinDistance = distance;
                index = i;
            }
        }
        List<Point> list = clusters.get(index);
        list.add(point);
        clusters.put(index, list);
    }

    //计算每个类簇的中心点，并返回中心点偏移误差
    public double getClusterCenterPoint(List<Point> centerPoints, HashMap<Integer, List<Point>> clusters){
        double error = 0;
        for (int i=0; i<centerPoints.size(); i++){
            Point tmpCenterPoint = centerPoints.get(i);
            double centerX = 0, centerY = 0;
            List<Point> lists = clusters.get(i);
            for (int j=0; j<lists.size(); j++){
                centerX += lists.get(j).getX();
                centerY += lists.get(j).getY();
            }
            centerX /= lists.size();
            centerY /= lists.size();
            error += Math.abs(centerX - tmpCenterPoint.getX());
            error += Math.abs(centerY - tmpCenterPoint.getY());
            centerPoints.set(i, new Point(centerX, centerY));
        }
        return error;
    }

    //计算点之间的距离，这里计算欧氏距离（不开方）
    public double calDistance(Point point1, Point point2){
        return Math.pow((point1.getX() - point2.getX()), 2) + Math.pow((point1.getY() - point2.getY()), 2);
    }

    //打印簇中心点坐标，及簇中其他点坐标
    public void show(List<Point> centerPoints, HashMap<Integer, List<Point>> clusters){
        for (int i=0; i<centerPoints.size(); i++){
            System.out.print(MessageFormat.format("类{0}的中心点: <{1}, {2}>",(i+1), centerPoints.get(i).getX(), centerPoints.get(i).getY()));
            List<Point> lists = clusters.get(i);
            System.out.print("\t类中成员点有：");
            for (int j=0; j<lists.size(); j++){
                System.out.print("<"+lists.get(j).getX()+ ","+ lists.get(j).getY()+">\t");
            }
            System.out.println();
        }
    }
}

File: PythonMethod/KMeans.py
# coding=utf-8
import math

from KMeansModel import Point

# KMeans聚类实现
class KMeans(object):
    def __init__(self, data, num):
        # 数据集中的点
        self.points = []
        # 簇中心点
        self.center_points = []
        # 聚类结果的集合簇
        self.clusters = {}
        # 类簇个数
        self.num = num
        # 读文件
        self.loadData(data)

    def loadData(self, dataset):
        # 读取数据集中的数据点
        for data in dataset:
            self.points.append(Point.Point(data[0], data[1]))
        # 初始化KMeans模型，选择前K个数据点作为簇中心点
        for i in range(self.num):
            self.center_points.append(self.points[i])
            self.clusters[i] = []

    def doKMeans(self):
        err = 100.0
        while err > 0.01*self.num:
            # 每次计算新簇时，清空原来聚类的簇
            for key in self.clusters:
                self.clusters[key] = []
            # 计算每个点所划分的簇
            for point in self.points:
                self.dispatchPointToCluster(point)
            # 计算每个簇的中心点，并得到更新后中心点的偏移误差
            err = self.getClusterCenterPoint()
            self.Show()
            print("******************")

    # 计算每个点所划分的簇
    def dispatchPointToCluster(self, point):
        index = 0
        min_distance = 100.0
        for i in range(len(self.center_points)):
            distance = self.getDistance(point, self.center_points[i])
            if distance < min_distance:
                min_distance = distance
                index = i
        point_list = self.clusters[index]
        point_list.append(point)
        self.clusters[index] = point_list

    # 计算每个簇的中心点，并得到更新后中心点的偏移误差
    def getClusterCenterPoint(self):
        err = 0
        for i in range(len(self.center_points)):
            tmpCenterPoint = self.center_points[i]
            centerX = 0
            centerY = 0
            point_list = self.clusters[i]
            for point in point_list:
                centerX += point.get_x()
                centerY += point.get_y()
            centerX /= len(point_list)
            centerY /= len(point_list)
            err += math.fabs(centerX - tmpCenterPoint.get_x())
            err += math.fabs(centerY - tmpCenterPoint.get_y())
            center = Point.Point(centerX, centerY)
            self.center_points[i] = center
        return err


    # 计算两点之间的距离
    def getDistance(self, point1, point2):
        return math.pow(point1.get_x()-point2.get_x(), 2) + math.pow(point1.get_y()-point2.get_y(), 2)


    def Show(self):
        for i in range(len(self.center_points)):
            print("Center point: ")
            self.center_points[i].show_point()
            print("cluster: ")
            for p in self.clusters[i]:
                p.show_point()

if __name__=='__main__':
    data = [[3,3],[4,10],[9,6],[14,8],[18,11],[21,7]]
    kmeans = KMeans(data, 3)
    kmeans.doKMeans()
